#Configfile
configfile: 'config.yaml'
GTF = config['GTF']
PICARD = config['PICARD']
REFFLAT = config['REFFLAT']
RRNAINTERVALS = config['RRNAINTERVALS']
TMPDIR = config['TMPDIR']

#subworkflow align:
    #snakefile: "star_align.snake"

#rule order
ruleorder: gene_count_paired > gene_count_single

rule all:
	input: expand('summary/{sample}', sample=config['Samples'])

rule rna_metrics:
    input: '{sample}.Aligned.bam'
    output: 'logs/{sample}_rna_metrics.txt'
    shell:
        """java -Xmx4000m -jar -Djava.io.tmpdir={TMPDIR} {PICARD} CollectRnaSeqMetrics\
        INPUT={input}\
        OUTPUT={output}\
        REF_FLAT={REFFLAT}\
        STRAND=NONE\
        RIBOSOMAL_INTERVALS={RRNAINTERVALS}
        """

rule gene_count_single:
	input:  sorted_sam = '{sample}.Aligned.bam',
            metrics = 'logs/{sample}_rna_metrics.txt'
	output: 'summary/{sample}'
	shell:
		"""htseq-count -f bam -i gene_name -m union -a 10 --stranded=no {input.sorted_sam} {GTF} > {output}"""

rule gene_count_paired:
    input:  sorted_sam = '{sample}.Aligned.bam',
            metrics = 'logs/{sample}_rna_metrics.txt',
            R2 = '{sample}_R2.fastq.gz'
    output: 'summary/{sample}'
    shell:
        """htseq-count -f bam -i gene_name -m union -a 10 --stranded=no -r name {input.sorted_sam} {GTF} > {output}"""