<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>heatmap &mdash; Counter RNAseq Window  documentation</title>
    
    <link rel="stylesheet" href="_static/style.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Counter RNAseq Window  documentation" href="index.html" />
    <link rel="next" title="argparse_util" href="argparse_util.html" />
    <link rel="prev" title="coverage" href="coverage.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="argparse_util.html" title="argparse_util"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="coverage.html" title="coverage"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Counter RNAseq Window  documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="heatmap">
<h1>heatmap<a class="headerlink" href="#heatmap" title="Permalink to this headline">¶</a></h1>
<p><a class="reference internal" href="#craw.heatmap.split_data" title="craw.heatmap.split_data"><code class="xref py py-func docutils literal"><span class="pre">craw.heatmap.split_data()</span></code></a> and  <a class="reference internal" href="#craw.heatmap.sort" title="craw.heatmap.sort"><code class="xref py py-func docutils literal"><span class="pre">craw.heatmap.sort()</span></code></a> work on data freshly parsed from coverage file.
That mean that the data contain the metadata (all columns which are not coverage scores
like chromosome, position strand , on so on)</p>
<p>The other functions sort normalization function work on pandas 2D DataFrame or numpy arrays containing only
scores of coverage. That means all metadata was removed (<a class="reference internal" href="#craw.heatmap.remove_metadata" title="craw.heatmap.remove_metadata"><code class="xref py py-func docutils literal"><span class="pre">craw.heatmap.remove_metadata()</span></code></a>).</p>
<div class="section" id="sort">
<h2>sort<a class="headerlink" href="#sort" title="Permalink to this headline">¶</a></h2>
<p>The is one public sort function which act as proxy for several private
sorting function.</p>
</div>
<div class="section" id="normalisation">
<h2>normalisation<a class="headerlink" href="#normalisation" title="Permalink to this headline">¶</a></h2>
<p>Several functions to normalize data.</p>
<p>The data can be normalize using min max of the whole data.
Or the min max is recalculated for each row.</p>
<p>in both case the formula is</p>
<p>zi = xi - min(x) / max(x) - min(x)</p>
<p>where x=(x1,...,xn) and zi is now your with normalized data.
in first case x is the whole matrix in 2nd is the row.</p>
<p>Normalization can be precede by 10 base log transformation.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In this case all 0 values are replace by 1 (10 base log is not define)</p>
</div>
</div>
<div class="section" id="drawing-heatmap">
<h2>drawing heatmap<a class="headerlink" href="#drawing-heatmap" title="Permalink to this headline">¶</a></h2>
<p>There are 2 way to generates figures, the first one is to generate a figures containing 2 heatmap for sense or antisense
with axis, legend on so on. But in this representation it&#8217;s not possible to display a figure with no scaling out/in.
So the information of one pixel is not accessible. This representation is generate by <a class="reference internal" href="#craw.heatmap.draw_heatmap" title="craw.heatmap.draw_heatmap"><code class="xref py py-func docutils literal"><span class="pre">craw.heatmap.draw_heatmap()</span></code></a>
and use matplotlib.</p>
<p>The second representation is to produce raw image where one nucleotide (one position for one gene) is represent by one pixel
without any scale in/out.
In this representation there si not axis legend on so on it&#8217;s only a raw image.</p>
<div class="section" id="heatmap-api-reference">
<h3>heatmap API reference<a class="headerlink" href="#heatmap-api-reference" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><span class="target" id="module-craw.heatmap"></span><dl class="class">
<dt id="craw.heatmap.Mark">
<em class="property">class </em><code class="descclassname">craw.heatmap.</code><code class="descname">Mark</code><span class="sig-paren">(</span><em>pos</em>, <em>data</em>, <em>color_map</em>, <em>color=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#Mark"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.Mark" title="Permalink to this definition">¶</a></dt>
<dd><p>A mark is a position and a color tight together. 
It is used to draw a colored vertical line at the given position on the heatmap</p>
<dl class="method">
<dt id="craw.heatmap.Mark.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>pos</em>, <em>data</em>, <em>color_map</em>, <em>color=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#Mark.__init__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.Mark.__init__" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>pos</strong> (<em>int</em>) &#8211; The position where to draw a mark, the position is relative to the reference position (0)</li>
<li><strong>data</strong> (<code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object) &#8211; the coverage matrix</li>
<li><strong>color_map</strong> (<em>class`:matplotlib.pyplot.ColorMap` object</em>) &#8211; the color map used to draw the heatmap</li>
<li><strong>color</strong> &#8211; the color of the line, the supported formats are
- hexadecimal values as #rgb or #rrggbb, for instance #ff0000 is pure red.
- common html color names</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="craw.heatmap.Mark.__weakref__">
<code class="descname">__weakref__</code><a class="headerlink" href="#craw.heatmap.Mark.__weakref__" title="Permalink to this definition">¶</a></dt>
<dd><p>list of weak references to the object (if defined)</p>
</dd></dl>

<dl class="method">
<dt id="craw.heatmap.Mark._color_converter">
<code class="descname">_color_converter</code><span class="sig-paren">(</span><em>color</em>, <em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#Mark._color_converter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.Mark._color_converter" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>color</strong> (<em>string</em>) &#8211; the color of the line, the supported formats are
- hexadecimal values as #rgb or #rrggbb, for instance #ff0000 is pure red.
- common html color names</li>
<li><strong>data</strong> (<code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object) &#8211; the matrix coverage</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">rgb color</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">tuple with 3 int between 0 and 255</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="craw.heatmap.Mark._get_matrix_bound">
<code class="descname">_get_matrix_bound</code><span class="sig-paren">(</span><em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#Mark._get_matrix_bound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.Mark._get_matrix_bound" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>data</strong> (<code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object) &#8211; the matrix coverage</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the most right and left position of the coverage</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">tuple of 2 int</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="craw.heatmap.Mark.to_px">
<code class="descname">to_px</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#Mark.to_px"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.Mark.to_px" title="Permalink to this definition">¶</a></dt>
<dd><p>tanslate the position of the mark relative to the reference in pixel.
:return: the position of the mark in pixel.
:rtype: positive int</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="craw.heatmap._sort_by_gene_size">
<code class="descclassname">craw.heatmap.</code><code class="descname">_sort_by_gene_size</code><span class="sig-paren">(</span><em>data</em>, <em>start_col=None</em>, <em>stop_col=None</em>, <em>ascending=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#_sort_by_gene_size"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap._sort_by_gene_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Sort the matrix in function of the gene size.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code>.) &#8211; the data to sort.</li>
<li><strong>start_col</strong> (<em>string.</em>) &#8211; the name of the column representing the beginning of the gene.</li>
<li><strong>stop_col</strong> (<em>string</em>) &#8211; the name of the column representing the end of the gene.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">sorted data.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap._sort_using_col">
<code class="descclassname">craw.heatmap.</code><code class="descname">_sort_using_col</code><span class="sig-paren">(</span><em>data</em>, <em>col=None</em>, <em>ascending=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#_sort_using_col"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap._sort_using_col" title="Permalink to this definition">¶</a></dt>
<dd><p>Sort the matrix in function of the column col</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code>.) &#8211; the data to sort.</li>
<li><strong>col</strong> (<em>string.</em>) &#8211; the name of the column to use for sorting the data.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">sorted data.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap._sort_using_file">
<code class="descclassname">craw.heatmap.</code><code class="descname">_sort_using_file</code><span class="sig-paren">(</span><em>data</em>, <em>file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#_sort_using_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap._sort_using_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Sort the matrix in function of file.
The file must have the following structure
the first line must be the name of the column
the following lines must be the values, one per line
each line starting by &#8216;#&#8217; will be ignore.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code>.) &#8211; the data to sort.</li>
<li><strong>file</strong> (<em>a file like object.</em>) &#8211; The file to use as guide to sort the data.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">sorted data.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.crop_matrix">
<code class="descclassname">craw.heatmap.</code><code class="descname">crop_matrix</code><span class="sig-paren">(</span><em>data</em>, <em>start_col</em>, <em>stop_col</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#crop_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.crop_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Crop matrix (remove columns). The resulting matrix will be [start_col, stop_col]</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.) &#8211; the data to sort.</li>
<li><strong>start_col</strong> (<em>string.</em>) &#8211; The name of the first column to keep.</li>
<li><strong>stop_col</strong> (<em>string.</em>) &#8211; The name of the last column to keep.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">sorted data.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object or None if data is None.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.draw_heatmap">
<code class="descclassname">craw.heatmap.</code><code class="descname">draw_heatmap</code><span class="sig-paren">(</span><em>sense</em>, <em>antisense</em>, <em>color_map=&lt;matplotlib.colors.LinearSegmentedColormap object&gt;</em>, <em>title=''</em>, <em>sense_on='top'</em>, <em>size=None</em>, <em>marks=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#draw_heatmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.draw_heatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a figure with subplot to represent the data as heat map.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>sense</strong> (a <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.) &#8211; the data normalized (xi in [0,1]) representing coverage on sense.</li>
<li><strong>antisense</strong> &#8211; the data normalized (xi in [0,1]) representing coverage on anti sense.</li>
<li><strong>color_map</strong> (a <code class="xref py py-class docutils literal"><span class="pre">matplotlib.pyplot.cm</span></code> object.) &#8211; the color map to use to represent the data.</li>
<li><strong>title</strong> (<em>string.</em>) &#8211; the figure title (by default the same as the coverage file).</li>
<li><strong>sense_on</strong> (<em>string.</em>) &#8211; specify the lay out. Where to place the heat map representing the sense data.
the available values are: &#8216;left&#8217;, &#8216;right&#8217;, &#8216;top&#8217;, &#8216;bottom&#8217; (default = &#8216;top&#8217;).</li>
<li><strong>size</strong> (<em>tuple of 2 float.</em>) &#8211; the size of the figure in inches (wide, height).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The figure.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a <code class="xref py py-class docutils literal"><span class="pre">matplotlib.pyplot.Figure</span></code> object.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.draw_one_matrix">
<code class="descclassname">craw.heatmap.</code><code class="descname">draw_one_matrix</code><span class="sig-paren">(</span><em>mat</em>, <em>ax</em>, <em>cmap=&lt;matplotlib.colors.LinearSegmentedColormap object&gt;</em>, <em>y_label=None</em>, <em>marks=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#draw_one_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.draw_one_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Draw a matrix using matplotlib imshow object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>mat</strong> (a <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.) &#8211; the data to represent graphically.</li>
<li><strong>ax</strong> (a <code class="xref py py-class docutils literal"><span class="pre">matplotlib.axis</span></code> object) &#8211; the axis where to represent the data</li>
<li><strong>cmap</strong> (a <code class="xref py py-class docutils literal"><span class="pre">matplotlib.pyplot.cm</span></code> object.) &#8211; the color map to use to represent the data.</li>
<li><strong>y_label</strong> (<em>string</em>) &#8211; the label for the data draw on y-axis.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the mtp image corresponding to data</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a <code class="xref py py-class docutils literal"><span class="pre">matplotlib.image</span></code> object.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.draw_raw_image">
<code class="descclassname">craw.heatmap.</code><code class="descname">draw_raw_image</code><span class="sig-paren">(</span><em>data</em>, <em>out_name</em>, <em>color_map=&lt;matplotlib.colors.LinearSegmentedColormap object&gt;</em>, <em>format='PNG'</em>, <em>marks=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#draw_raw_image"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.draw_raw_image" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate an image file with one pixel for each values of the data matrix.
the data can be either the coverage on sense or on antisense.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> or <code class="xref py py-class docutils literal"><span class="pre">numpy.array</span></code> object) &#8211; a <strong>Normalized</strong> (where all values are between 0 and 1) matrix.</li>
<li><strong>out_name</strong> (<em>string</em>) &#8211; The name of the generated graphic file.</li>
<li><strong>color_map</strong> &#8211; </li>
<li><strong>format</strong> (<em>string</em>) &#8211; the format of the result png, jpeg, ... (see pillow supported formats)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raise:</th><td class="field-body"><p class="first last">RuntimeError if data are not normalized.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.get_data">
<code class="descclassname">craw.heatmap.</code><code class="descname">get_data</code><span class="sig-paren">(</span><em>coverage_file</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#get_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.get_data" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>coverage_file</strong> (<em>str</em>) &#8211; the path of the coverage file to parse.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the data as 2 dimension dataframe</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">a <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.lin_norm">
<code class="descclassname">craw.heatmap.</code><code class="descname">lin_norm</code><span class="sig-paren">(</span><em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#lin_norm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.lin_norm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize data with linear algorithm.
The formula applied to obtain the results is:</p>
<blockquote>
<div>zi = xi - min(x) / max(x) - min(x)</div></blockquote>
<p>where x=(x1,...,xn) and zi is now your with normalized data.
Ensure that the resulting values are comprise between 0 and 1.
return None if data is None, return empty <code class="xref py py-class docutils literal"><span class="pre">pd.DataFrame</span></code> object if data is empty.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>data</strong> (a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.) &#8211; the data to normalize, this 2D matrix must contains only coverage scores (no more metadata).</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a normalize matrix, where  0 &lt;= zi &lt;=1 where z=(z1, ..., zn)</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object or None if data is None.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.lin_norm_row_by_row">
<code class="descclassname">craw.heatmap.</code><code class="descname">lin_norm_row_by_row</code><span class="sig-paren">(</span><em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#lin_norm_row_by_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.lin_norm_row_by_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize data with linear algorithm but
instead to normalize all the matrix, the normalization formula
(see <code class="xref py py-func docutils literal"><span class="pre">normalize()</span></code>) is applied row by row.
It ensure that all values are between 0 and 1.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>data</strong> (a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.) &#8211; the data to normalize, this 2D matrix must contains only coverage scores (no more metadata).</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a normalize matrix, where  0 &lt;= zi &lt;=1 where z=(z1, ..., zn)</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object or None if data is None.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.log_norm">
<code class="descclassname">craw.heatmap.</code><code class="descname">log_norm</code><span class="sig-paren">(</span><em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#log_norm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.log_norm" title="Permalink to this definition">¶</a></dt>
<dd><p>The base 10 logarithm is compute for all values before
a normalization (see <code class="xref py py-func docutils literal"><span class="pre">normalize()</span></code> ) to ensure
that all values are comprise between 0 and 1 .</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">coverage scores are integers &gt;= 0.
log10(0) = -inf or warning in macos
prior to normalize data the 0 values are replace by 1.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>data</strong> (a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.) &#8211; the data to normalize, this 2D matrix must contains only coverage scores (no more metadata).</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a normalize matrix, where  0 &lt;= zi &lt;=1 where z=(z1, ..., zn)</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object or None if data is None.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.log_norm_row_by_row">
<code class="descclassname">craw.heatmap.</code><code class="descname">log_norm_row_by_row</code><span class="sig-paren">(</span><em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#log_norm_row_by_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.log_norm_row_by_row" title="Permalink to this definition">¶</a></dt>
<dd><p>as <code class="xref py py-func docutils literal"><span class="pre">normalize_row_by_row()</span></code> but prior normalisation
a 10 base logarithm is applied.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">coverage scores are integers &gt;= 0.
log10(0) = -inf
to normalize data the -inf value are change in 0.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>data</strong> (a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.) &#8211; the data to normalize, this 2D matrix must contains only coverage scores (no more metadata).</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a normalize matrix, where  0 &lt;= zi &lt;=1 where z=(z1, ..., zn)</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object or None if data is None.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.remove_metadata">
<code class="descclassname">craw.heatmap.</code><code class="descname">remove_metadata</code><span class="sig-paren">(</span><em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#remove_metadata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.remove_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove all information which is not coverage value (as chromosome, strand, name, ...)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>data</strong> (<code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code>.) &#8211; the data coming from a coverage file parsing containing coverage information and metadata chromosome,
gene name , ...</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">sorted data.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">a 2D <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object or None if data is None.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.sort">
<code class="descclassname">craw.heatmap.</code><code class="descname">sort</code><span class="sig-paren">(</span><em>data</em>, <em>criteria</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#sort"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.sort" title="Permalink to this definition">¶</a></dt>
<dd><p>Sort the matrix in function of criteria.
This function act as proxy for several specific sorting functions</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code>.) &#8211; the data to sort.</li>
<li><strong>criteria</strong> (<em>string.</em>) &#8211; which criteria to use to sort the data (by_gene_size, using_col, using_file).</li>
<li><strong>kwargs</strong> &#8211; depending of the criteria
- start_col, stop_col for sort_by_gene_size
- col for using_col
- file for using file</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">sorted data.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="craw.heatmap.split_data">
<code class="descclassname">craw.heatmap.</code><code class="descname">split_data</code><span class="sig-paren">(</span><em>data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/heatmap.html#split_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.heatmap.split_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Split the matrix in 2 matrices one for sense the other for antisense.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>data</strong> (a 2 dimension <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object) &#8211; the coverage data to split</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">two matrix</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">tuple of two <code class="xref py py-class docutils literal"><span class="pre">pandas.DataFrame</span></code> object (sense pandas.DataFrame, antisense pandas.DataFrame)</td>
</tr>
</tbody>
</table>
</dd></dl>

</div></blockquote>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">heatmap</a><ul>
<li><a class="reference internal" href="#sort">sort</a></li>
<li><a class="reference internal" href="#normalisation">normalisation</a></li>
<li><a class="reference internal" href="#drawing-heatmap">drawing heatmap</a><ul>
<li><a class="reference internal" href="#heatmap-api-reference">heatmap API reference</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="coverage.html"
                        title="previous chapter">coverage</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="argparse_util.html"
                        title="next chapter">argparse_util</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/heatmap.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="argparse_util.html" title="argparse_util"
             >next</a> |</li>
        <li class="right" >
          <a href="coverage.html" title="coverage"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Counter RNAseq Window  documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2017, Bertrand Néron.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>