<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Inputs / Outputs &mdash; Counter RNAseq Window  documentation</title>
    
    <link rel="stylesheet" href="_static/style.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Counter RNAseq Window  documentation" href="index.html" />
    <link rel="next" title="craw" href="craw.html" />
    <link rel="prev" title="Quick start" href="quickstart.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="craw.html" title="craw"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="quickstart.html" title="Quick start"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Counter RNAseq Window  documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="inputs-outputs">
<span id="id1"></span><h1>Inputs / Outputs<a class="headerlink" href="#inputs-outputs" title="Permalink to this headline">¶</a></h1>
<div class="section" id="craw-coverage">
<h2>craw_coverage<a class="headerlink" href="#craw-coverage" title="Permalink to this headline">¶</a></h2>
<div class="section" id="inputs">
<h3>Inputs<a class="headerlink" href="#inputs" title="Permalink to this headline">¶</a></h3>
<p>craw_coverage* need a file bam or wig to compute coverage and an annotation file
to specify on which regions to compute these coverages.</p>
<div class="section" id="bam-file">
<h4>bam file<a class="headerlink" href="#bam-file" title="Permalink to this headline">¶</a></h4>
<p><em>craw_coverage</em> can use a file of alignment reads called bam file.
a bam file is a short DNA sequence read alignments in the Binary Alignment/Map format (.bam).
<em>craw_coverage</em> needs also the corresponding index file (bai). The index file must be located beside the bam file
with the same name instead to have the <em>.bam</em> extension it end by <em>.bai</em> extension.
If you have not  the index file you have to create it.</p>
<p>To index a bam file you need samtools. The command line is</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>samtools index file.bam
</pre></div>
</div>
<p>For more explanation see <a class="reference external" href="http://www.htslib.org/doc/">http://www.htslib.org/doc/</a> .</p>
</div>
<div class="section" id="wig-file">
<h4>wig file<a class="headerlink" href="#wig-file" title="Permalink to this headline">¶</a></h4>
<p><em>craw_coverage</em> can compute coverage also from wig file
see <a class="reference external" href="https://wiki.nci.nih.gov/display/tcga/wiggle+format+specification">https://wiki.nci.nih.gov/display/tcga/wiggle+format+specification</a> and
<a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">http://genome.ucsc.edu/goldenPath/help/wiggle.html</a> .
for format specifications. Compare d to these specifications
craw support coverages on both strands. the positive coverages scores
are on the forward strand whereas the negative ones are on the reverse strand.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>track type=wiggle_0 name=&quot;demo&quot; color=96,144,246 altColor=96,144,246 autoScale=on  graphType=bar
variableStep chrom=chrI span=1
72      12.0000
73      35.0000
74      70.0000
75      127.0000
...
72      -88.0000
73      -42.0000
74      -12.0000
75      -1.0000
</pre></div>
</div>
<p>In the example above the coverage on the Chromosome I for the positions 72, 73, 74, 75
are 12, 35, 70, 127 on the forward strand and 88, 42, 12, 1 on the reverse strand.</p>
</div>
<div class="section" id="annotation-file">
<h4>annotation file<a class="headerlink" href="#annotation-file" title="Permalink to this headline">¶</a></h4>
<p>The annotation file is a <cite>tsv</cite> file by default.
It&#8217;s mean that it is a text file with value separated by tabulation (not spaces) or commas.
But if a  separator is specified (&#8211;sep) it can be a csv file or any columns file.</p>
<p>The first line of the file must be the name of the columns
the other lines the values. Each line represent a row.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>name    gene    chromosome      strand  Position
YEL072W RMD6    chrV    +       14415
YEL071W DLD3    chrV    +       17845
YEL070W DSF1    chrV    +       21097
YEL066W HPA3    chrV    +       27206
YEL065W SIT1    chrV    +       29543
YEL062W NPR2    chrV    +       36254
YEL058W PCM1    chrV    +       44925
YEL056W HAT2    chrV    +       48373
</pre></div>
</div>
<p>All lines starting with &#8216;#&#8217; character will be ignored.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span># This is the annotation file for Wild type
# bla bla ...
name    gene    chromosome      strand  Position
YEL072W RMD6    chrV    +       14415
YEL071W DLD3    chrV    +       17845
YEL070W DSF1    chrV    +       21097
YEL066W HPA3    chrV    +       27206
YEL065W SIT1    chrV    +       29543
YEL062W NPR2    chrV    +       36254
YEL058W PCM1    chrV    +       44925
YEL056W HAT2    chrV    +       48373
</pre></div>
</div>
<div class="section" id="mandatory-columns">
<h5>mandatory columns<a class="headerlink" href="#mandatory-columns" title="Permalink to this headline">¶</a></h5>
<p>There is 3 mandatory columns in the annotation file.</p>
<div class="section" id="columns-with-fixed-name">
<h6>columns with fixed name<a class="headerlink" href="#columns-with-fixed-name" title="Permalink to this headline">¶</a></h6>
<p>two with a fixed name:</p>
<ul class="simple">
<li><strong>strand</strong> indicate on which strand is located the region of interest. The authorized values for this columns are +/- , 1/-1 or for/rev.</li>
<li><strong>chromosome</strong> the chromosome name where is located the region of interest.</li>
</ul>
</div>
<div class="section" id="columns-with-variable-name">
<h6>columns with variable name<a class="headerlink" href="#columns-with-variable-name" title="Permalink to this headline">¶</a></h6>
<p>In addition of these two columns the column to define the position of reference is mandatory too, but the name of this
column can be specified by the user. If it&#8217;s not craw_coverage will use a column name &#8216;position&#8217;.</p>
<p>If we want to compute coverage on variable window size, 2 extra columns whose name must be specified by the user by the following option:</p>
<ul class="simple">
<li>&#8211;start-col to define the beginning of the window (this position is included in the window)</li>
<li>&#8211;stop-col to define the end of the window (this position is included in the window)</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre><span></span>name    gene    type    chromosome      strand  annotation_start        annotation_end  has_transcript  transcription_end       transcription_start
YEL072W RMD6    gene    chrV    1       13720   14415   1       14745   13569
YEL071W DLD3    gene    chrV    1       16355   17845   1       17881   16177
YEL070W DSF1    gene    chrV    1       19589   21097   1       21197   19539
YEL066W HPA3    gene    chrV    1       26721   27206   1       27625   26137
YEL065W SIT1    gene    chrV    1       27657   29543   1       29601   27625
YEL062W NPR2    gene    chrV    1       34407   36254   1       36401   34321
YEL058W PCM1    gene    chrV    1       43252   44925   1       44993   43217
YEL056W HAT2    gene    chrV    1       47168   48373   1       48457   47105
YEL052W AFG1    gene    chrV    1       56571   58100   1       58105   56537
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span>craw_coverage --wig file.wig --annot annot.txt --ref-col annotation_start --start-col annotation_start --stop-col annotation_end
</pre></div>
</div>
<p>The position of reference must be between start and end.
The authorized values are positive integers.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>the position of reference can be used to define the reference and the start ot the end of the window.</p>
<div class="last highlight-python"><div class="highlight"><pre><span></span>craw_coverage --bam file.bam --annot annot.txt --ref-col annotation_start --start-col annotation_start --stop-col annotation_end
</pre></div>
</div>
</div>
<p>All other columns are not necessary but will be reported as is in the coverage file.</p>
</div>
</div>
</div>
</div>
<div class="section" id="outputs">
<span id="cov-out"></span><h3>Outputs<a class="headerlink" href="#outputs" title="Permalink to this headline">¶</a></h3>
<div class="section" id="coverage-file">
<h4>coverage_file<a class="headerlink" href="#coverage-file" title="Permalink to this headline">¶</a></h4>
<p>It&#8217;s a <cite>tsv</cite> file with all columns found in annotation file plus the result of coverage position by position centered
on the reference position define for each line. for instance</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>craw_coverage --wig=../data/small.wig --annot=../data/annotations.txt
--ref-col=annotation_start --before=0  --after=2000
</pre></div>
</div>
<p>In the command line above, the column &#8216;0&#8217; correspond to the annotation_start position the column &#8216;1&#8217; to annotation_start + 1
on so on until &#8216;2000&#8217; (here we display only the first 3 columns of the coverage).</p>
<div class="highlight-python"><div class="highlight"><pre><span></span># Running Counter RnAseq Window craw_coverage
# Version: craw NOT packaged, it should be a development version | Python 3.4
# Using: pysam 0.9.1.4 (samtools 1.3.1)
#
# craw_coverage run with the following arguments:
# --after=3
# --annot=../data/annotation_wo_start.txt
# --before=5
# --chr-col=chromosome
# --output=small_wig.cov
# --qual-thr=0
# --quiet=1
# --ref-col=Position
# --sense=mixed
# --sep=
# --strand-col=strand
# --suffix=cov
# --verbose=0
# --wig=../data/small.wig
sense   name    gene    type    chromosome      strand  annotation_start        annotation_end  has_transcript  transcription_end       transcription_start     0       1       2
S       YEL072W RMD6    gene    chrV    +       13720   14415   1       14745   13569   7       7       7
AS      YEL072W RMD6    gene    chrV    +       13720   14415   1       14745   13569   0       0       0
S       YEL071W DLD3    gene    chrV    +       16355   17845   1       17881   16177   31      33      33
</pre></div>
</div>
<p>The line starting with &#8216;#&#8217; are comments and will be ignored for further processing.
But in traceability/reproducibility concern, in the comments <cite>craw_coverage</cite> indicate
the version of the program and the arguments used for this experiment.</p>
</div>
</div>
</div>
<div class="section" id="craw-htmp">
<h2>craw_htmp<a class="headerlink" href="#craw-htmp" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id2">
<h3>Inputs<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>see <a class="reference internal" href="#cov-out">cov_out</a></p>
</div>
<div class="section" id="id3">
<h3>Outputs<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>The default output of <em>craw_htmp</em> (if &#8211;out is omitted) is grapical window on the screen.
The figure display on the screen can be saved using the window menu.</p>
<div class="container">
<a class="reference internal image-reference" href="_images/craw_htmp_sense_on_left.png"><img alt="--sense-on-left" src="_images/craw_htmp_sense_on_left.png" style="width: 30%;" /></a>
</div>
<p>It is also possible to generate directly a image file in various format by specifying the &#8211;out option.
The output format will be deduced form the filename extension provide to &#8211;out option.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>--out foo.jpeg  for jpeg image or --out foo.png  for png image
</pre></div>
</div>
<p>The supported format vary in function of the matloblib backend used (see <a class="reference internal" href="installation.html#matplotlibrc"><span>matplotlib configuration</span></a>).</p>
<p>If &#8211;size raw is used 2 files will be generated one for the sense and the other for the antisense.
If &#8211;out is not specified it will be the name of the coverage file without extension and the format will be png.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>craw_htmp foo_bar.cov --size raw
</pre></div>
</div>
<p>will produce <em>foo_bar.sense.png</em> and <em>foo_bar.antisense.png</em></p>
<div class="highlight-python"><div class="highlight"><pre><span></span>craw_htmp foo_bar.cov --size raw --out Xyzzy.jpeg
</pre></div>
</div>
<p>will produce <em>Xyzzy.sense.jpeg</em> and <em>Xyzzy.antisense.jpeg</em></p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Inputs / Outputs</a><ul>
<li><a class="reference internal" href="#craw-coverage">craw_coverage</a><ul>
<li><a class="reference internal" href="#inputs">Inputs</a><ul>
<li><a class="reference internal" href="#bam-file">bam file</a></li>
<li><a class="reference internal" href="#wig-file">wig file</a></li>
<li><a class="reference internal" href="#annotation-file">annotation file</a><ul>
<li><a class="reference internal" href="#mandatory-columns">mandatory columns</a><ul>
<li><a class="reference internal" href="#columns-with-fixed-name">columns with fixed name</a></li>
<li><a class="reference internal" href="#columns-with-variable-name">columns with variable name</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#outputs">Outputs</a><ul>
<li><a class="reference internal" href="#coverage-file">coverage_file</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#craw-htmp">craw_htmp</a><ul>
<li><a class="reference internal" href="#id2">Inputs</a></li>
<li><a class="reference internal" href="#id3">Outputs</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="quickstart.html"
                        title="previous chapter">Quick start</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="craw.html"
                        title="next chapter">craw</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/inputs-outputs.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="craw.html" title="craw"
             >next</a> |</li>
        <li class="right" >
          <a href="quickstart.html" title="Quick start"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Counter RNAseq Window  documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2017, Bertrand Néron.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>