#Configfile
configfile: 'config.json'

#Executables
TMPDIR = config['TMPDIR']
PICARD = config['PICARD']
DROPSEQ = config['DROPSEQ']

#References files
GENOMEREF = config['GENOMEREF']
REFFLAT = config['REFFLAT']
GTF = config['GTF']



rule all:
	input: expand('{sample}_hist_out_cell.txt', sample=config['Samples'])
	log: 'logs/Dropseq_post_align.log'

rule sort:
	input: 
		samples = '{sample}_Aligned.sam'
	output: '{sample}_Aligned_sorted.sam'
	threads:1
	shell:
		"""java	-Djava.io.tmpdir={TMPDIR} -Dsamjdk.buffer_size=131072 -XX:GCTimeLimit=50 -XX:GCHeapFreeLimit=10 -Xmx4000m -jar {PICARD} SortSam\
		INPUT={input}\
		OUTPUT={output}\
		SORT_ORDER=queryname\
		TMP_DIR={TMPDIR}"""

rule stage3:
	input:	unmapped = '{sample}_tagged_unmapped.bam',
			mapped = '{sample}_Aligned_sorted.sam'
	output: '{sample}_gene_exon_tagged.bam'
	threads:2
	shell:
		"""java -Djava.io.tmpdir={TMPDIR} -Xmx4000m -jar {PICARD} MergeBamAlignment\
		REFERENCE_SEQUENCE={GENOMEREF}\
		UNMAPPED_BAM={input.unmapped}\
		ALIGNED_BAM={input.mapped}\
		INCLUDE_SECONDARY_ALIGNMENTS=false\
		PAIRED_RUN=false\
		OUTPUT=/dev/stdout COMPRESSION_LEVEL=0|\
		\
		{DROPSEQ}/TagReadWithGeneExon\
		O={output}\
		I=/dev/stdin\
		ANNOTATIONS_FILE={GTF}\
		TAG=GE\
		CREATE_INDEX=true
		"""

rule bead_errors:
	input: '{sample}_gene_exon_tagged.bam'
	output: '{sample}_final.bam'
	params:
		out_stats = 'summary/{sample}_synthesis_stats_summary.txt',
		summary = 'summary/{sample}_synthesis_stats_summary.txt',
		primer = lambda wildcards: config['Primers'][config['Samples'][wildcards.sample]],
		barcodes = config['Barcodes']
	shell:
		"""{DROPSEQ}/DetectBeadSynthesisErrors\
		INPUT={input}\
		OUTPUT={output}\
		OUTPUT_STATS={params.out_stats}\
		SUMMARY={params.summary}\
		NUM_BARCODES={params.barcodes}\
		PRIMER_SEQUENCE=CAAGCAGAAGACGGCATACGAGAT{params.primer}GTCTCGTGGGCTCGG
		"""

rule bam_hist:
	input: '{sample}_final.bam'
	output: '{sample}_hist_out_cell.txt'
	shell:
		"""{DROPSEQ}/BAMTagHistogram\
		I={input}\
		O={output}\
		TAG=XC
		"""
