#Configfile
configfile:'config.yaml'

STAREXEC = config['STAREXEC']
GTF = config['GTF']
METAREF = config['METAREF']
MISMATCH = config['GLOBAL']['allowed_aligner_mismatch']

#subworkflow align:
    #snakefile: "pre_align.snake"

#rule order
ruleorder: STAR_align_paired > STAR_align_single

rule all:
	input: expand('{sample}.Aligned.bam', sample=config['Samples'])
	shell:"""{STAREXEC}\
		--genomeDir {METAREF}\
		--genomeLoad Remove;\
		rm -rf Aligned.sortedByCoord.out.bam Log.out Log.progress.out _STARtmp"""


rule STAR_align_paired:
	input: 
		R1 = temp('{sample}_trimmed_R1.fastq.gz'),
		R2 = temp('{sample}_trimmed_R2.fastq.gz')
	output: 'logs/{sample}.Aligned.sortedByCoord.out.bam'
	threads: 8
	params:
		prefix = '{sample}.',
		mismatch = MISMATCH
	shell:"""{STAREXEC}\
			--genomeDir {METAREF}\
			--runThreadN 8\
			--outFilterMismatchNmax={params.mismatch}\
			--readFilesIn {input.R1} {input.R2}\
			--genomeLoad LoadAndKeep\
			--quantMode GeneCounts\
			--readFilesCommand zcat\
			--outFileNamePrefix logs/{params.prefix}\
			--outSAMtype BAM SortedByCoordinate\
			--limitBAMsortRAM 10000000000"""

rule STAR_align_single:
	input: temp('{sample}_trimmed_R1.fastq.gz')
	output: 'logs/{sample}.Aligned.sortedByCoord.out.bam'
	threads: 8
	params:
		prefix = '{sample}.',
		mismatch = MISMATCH
	shell:"""{STAREXEC}\
			--genomeDir {METAREF}\
			--runThreadN 8\
			--outFilterMismatchNmax={params.mismatch}\
			--readFilesIn {input}\
			--genomeLoad LoadAndKeep\
			--quantMode GeneCounts\
			--readFilesCommand zcat\
			--outFileNamePrefix logs/{params.prefix}\
			--outSAMtype BAM SortedByCoordinate\
			--limitBAMsortRAM 10000000000"""


rule mv_sam:
	input: 'logs/{sample}.Aligned.sortedByCoord.out.bam'
	output: '{sample}.Aligned.bam'
	params: prefix = '{sample}'
	shell:"""mv {input} {output}; mv logs/{params.prefix}.ReadsPerGene.out.tab summary/{params.prefix}.ReadsPerGene.out.tab"""

